src/rpp/rpp/operators/take_until.hpp

```diff
@@ -0,0 +1,108 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2022 - present.
+//                    TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/defs.hpp>
+#include <rpp/operators/fwd/take_until.hpp>
+#include <rpp/operators/first.hpp>
+#include <rpp/sources/create.hpp>
+#include <rpp/subscribers/constraints.hpp>
+#include <rpp/utils/functors.hpp>
+
+#include <rpp/operators/details/subscriber_with_state.hpp> // create_subscriber_with_state
+
+#include <mutex>
+
+IMPLEMENTATION_FILE(take_until_tag);
+
+namespace rpp::details
+{
+
+struct take_until_state
+{
+    std::mutex mutex;
+    bool is_stopped{false};
+};
+
+/**
+ * Functor (type-erasure) of "take_until" for on_next operator.
+ */
+template<constraint::decayed_type Type>
+struct take_until_on_next
+{
+    void operator()(auto&& value,
+                    const auto& subscriber,
+                    const std::shared_ptr<take_until_state>& state) const
+    {
+        std::lock_guard lock{state->mutex};
+
+        if (!state->is_stopped)
+            subscriber.on_next(std::forward<decltype(value)>(value));
+        else
+            subscriber.on_completed();
```

Comment:
Looks like this line useless. If `is_stopped` true, then we've send `on_completed`/`on_error` from another observable